### PR TITLE
Reset outgoing stream on inbound is reset

### PR DESCRIPTION
**Pull request by enobufs**
Added state to Stream to reset outbound only when the state is open Relates to #187

Hi @jerry-tao, what do you think about this fix. (this branch is based on your fork `jerry-tao/sctp:master`).

Who to react to the incoming OutgoingResetRequest should be Stream layer (not at the Association layer).

Also, if the stream already sent its OutgoingResetRequest, it shouldn't send another one, meaning, we need to introduce a state in the Stream class (s.state), and send OutgoingResetRequest if the state is "open", I believe. Which prevented the vnet test case from failing.

As [RFC 8831](https://www.rfc-editor.org/rfc/rfc8831.html#name-closing-a-data-channel) implies, teardown sequence of a stream starts with outgoing (closing), then incoming (closed). This branch also addresses it by not to cause error (s.readErr) because it will eventually be reset by the remote.

I am hoping to merge this into your branch, then we can finalize it in your pull-request.

### CODE REVIEW

File: association.go [hunk 1]
```diff
@@ -2023,41 +2026,12 @@ func (a *Association) resetStreamsIfAny(p *paramOutgoingResetRequest) *packet {
...
-		&chunkReconfig{
-			paramA: &paramReconfigResponse{
-				reconfigResponseSequenceNumber: p.reconfigRequestSequenceNumber,
-				result:                         result,
-			},
+			result:                         result,
 		},
-		c,
-	})
-
```
Comment (enobufs) [inline association.go:2060]: Here, I am simply reverting the change.

File: association.go [hunk 2]
```diff
@@ -2014,6 +2014,9 @@ func (a *Association) resetStreamsIfAny(p *paramOutgoingResetRequest) *packet {
 			if !ok {
 				continue
 			}
+			a.lock.Unlock()
+			s.onInboundStreamReset()
+			a.lock.Lock()
```
Comment (enobufs) [inline association.go:2019]: Reporting the event to the corresponding stream and let it decide what to do.

File: stream.go [hunk 3]
```diff
@@ -178,32 +196,19 @@ func (s *Stream) Write(p []byte) (n int, err error) {
...
 
-	switch s.association.getState() {
-	case shutdownSent, shutdownAckSent, shutdownPending, shutdownReceived:
-		s.lock.Lock()
-		if s.writeErr == nil {
-			s.writeErr = errStreamClosed
-		}
-		s.lock.Unlock()
-	default:
-	}
```
Comment (enobufs) [inline stream.go:195]: This state checks are done inside `sendPayloadData()`

File: stream.go [hunk 4]
```diff
@@ -178,32 +196,19 @@ func (s *Stream) Write(p []byte) (n int, err error) {
...
-
-	s.lock.RLock()
-	err = s.writeErr
-	s.lock.RUnlock()
+	chunks := s.packetize(p, ppi)
+	n := len(p)
+	err := s.association.sendPayloadData(chunks)
 	if err != nil {
-		return 0, err
+		return n, errStreamClosed
```
Comment (enobufs) [inline stream.go:209]: Try to use the same error object to be compatible.

File: stream.go [hunk 5]
```diff
@@ -267,26 +272,18 @@ func (s *Stream) packetize(raw []byte, ppi PayloadProtocolIdentifier) []*chunkPa
...
-		isOpen := true
-		if s.writeErr == nil {
-			s.writeErr = errStreamClosed
-		} else {
-			isOpen = false
-		}
+		s.log.Debugf("[%s] Close: state=%s", s.name, s.state.String())
 
-		if s.readErr == nil {
-			s.readErr = io.EOF
```
Comment (enobufs) [inline stream.go:282]: This is not what RFC 8831 implies. s.Close() should only close the outgoing stream. (fixing the existing bug)

File: stream.go [hunk 6]
```diff
@@ -365,3 +362,25 @@ func (s *Stream) getNumBytesInReassemblyQueue() int {
...
+
+	// From draft-ietf-rtcweb-data-channel-13 section-6.7:
+	//	if one side decides to close the data channel, it resets the corresponding
+	//	outgoing stream.  When the peer sees that an incoming stream was
+	//	reset, it also resets its corresponding outgoing stream.  Once this
+	//	is completed, the data channel is closed.
+	if resetOutbound {
+		s.association.sendResetRequest(sid)
+	}
+}
```
Comment (enobufs) [inline stream.go:386]: Finally, this is what we wanted in the PR.

**Comment by jerry-tao** [comment]
@enobufs Sorry for the delay, just come back form vacation.

**Pull request merged**